typescript/sdk/src/providers/transactions/submitter/builder/TxSubmitterBuilder.ts

```diff
@@ -0,0 +1,163 @@
+import assert from 'assert';
+import { Logger } from 'pino';
+
+import { rootLogger } from '@hyperlane-xyz/utils';
+
+import { MultiProvider } from '../../../MultiProvider.js';
+import { HyperlaneTx } from '../../HyperlaneTx.js';
+import { HyperlaneTxReceipt } from '../../HyperlaneTxReceipt.js';
+import { InterchainAccountTxTransformer } from '../../transformer/InterchainAccountTxTransformer.js';
+import { TxTransformerInterface } from '../../transformer/TxTransformer.js';
+import { TxTransformerType } from '../../transformer/TxTransformerTypes.js';
+import { GnosisSafeTxSubmitter } from '../GnosisSafeTxSubmitter.js';
+import { ImpersonatedAccountTxSubmitter } from '../ImpersonatedAccountTxSubmitter.js';
+import { SignerTxSubmitter } from '../SignerTxSubmitter.js';
+import { TxSubmitterInterface } from '../TxSubmitter.js';
+import { TxSubmitterType } from '../TxSubmitterTypes.js';
+
+import {
+  TxSubmitterXORType,
+  TxTransformerXORType,
+} from './TxSubmitterBuilderTypes.js';
+
+/**
+ * Builds a TxSubmitterBuilder for batch transaction submission.
+ *
+ * Example use-cases:
+ *  const builder = new TxSubmitterBuilder(mp);
+ *  let txReceipts = builder.add(
+ *    TxSubmitterType.GNOSIS_SAFE, chainA
+ *  ).transform(
+ *    TxSubmitterType.ICA, chainB
+ *  ).submit(
+ *    populatedTxs
+ *  );
+ *  txReceipts = builder.add(new ImpersonatedAccountTxSubmitter(chainA)).submit(populatedTxs);
+ *  txReceipts = builder.add(TxSubmitterType.SIGNER, chainC).submit(populatedTxs);
+ */
+export class TxSubmitterBuilder<
+  HTX extends HyperlaneTx,
+  HTR extends HyperlaneTxReceipt,
+> {
+  protected readonly logger: Logger = rootLogger.child({
+    module: 'transactions',
+  });
+
+  private currentSubmitter?: TxSubmitterInterface<HTX, HTR>;
+  private currentTransformer?: TxTransformerInterface<HTX>;
```

Comment:
I would expect the builder to be fairly stateless

Reply:
Hmm– I'm not sure how we'd avoid maintaining state. If we'd like to have 1 or more transformers, we must keep track of what transformers have augmented the original txn set

Note: The transformer stack datastructure is now present

Reply:
```typescript
import { PopulatedTransaction } from "ethers";

interface TxSubmitter<ReturnType> {
    submit(txs: PopulatedTransaction[]): ReturnType;
}

type TxTransformer = TxSubmitter<PopulatedTransaction[]>;

const NoopTransformer: TxTransformer = {
  submit: (txs: PopulatedTransaction[]) => txs
}

class SubmitterBuilder<ReturnType> implements TxSubmitter<ReturnType> {
    constructor(
        private readonly inner: TxSubmitter<ReturnType>,
        private readonly transformer: TxTransformer = NoopTransformer
    ) {}

    add(transformer: TxTransformer): SubmitterBuilder<ReturnType> {
        return new SubmitterBuilder(this, transformer);
    }

    submit(txs: PopulatedTransaction[]): ReturnType {
        return this.inner.submit(this.transformer.submit(txs));
    }
}
```
already shared this but for posterity